kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/slot-updates.k

```diff
@@ -0,0 +1,144 @@
+    rule [gmw-bits-fail]:   #getMaskWidthBits(X:Int) => -1                                            requires -1 ==Int #getMaskShiftBits(X) [concrete(X), simplification]
+
+    rule [gmw-bytes-succ]: #getMaskWidthBytes(X:Int) => #getMaskWidthBits(X) /Int 8 requires           #getMaskWidthBits(X) modInt 8 ==Int 0   [concrete(X), simplification, preserves-definedness]
+    rule [gmw-bytes-fail]: #getMaskWidthBytes(X:Int) => -1                          requires notBool ( #getMaskWidthBits(X) modInt 8 ==Int 0 ) [concrete(X), simplification, preserves-definedness]
+
+    // Mask recogniser
+    syntax Bool ::= #isMask(Int) [function, total]
+
+    // A number is a mask if it has a valid shift, and a valid width, and all remaining bits set to one
+    rule [is-mask-true]:
+      #isMask(X:Int) => maxUInt256 ==Int X |Int ( 2 ^Int ( #getMaskShiftBits(X) +Int #getMaskWidthBits(X) ) -Int 1 )
+        requires 0 <=Int #getMaskShiftBytes(X) andBool 0 <=Int #getMaskWidthBytes(X)
+        [concrete(X), simplification, preserves-definedness]
+
+    // and is not a mask otherwise
+    rule [is-mask-false]:
+      #isMask(X:Int) => false
+        requires notBool ( 0 <=Int #getMaskShiftBytes(X) andBool 0 <=Int #getMaskWidthBytes(X) )
+        [concrete(X), simplification, preserves-definedness]
+
+    // ###########################################################################
+    // Masking lemmas
+    // ###########################################################################
+
+    // Slot updates are of the general form (SHIFT *Int VALUE) |Int (MASK &Int #asWord( SLOT:Bytes )),
+    // where the masking clears the part of the slot into which VALUE will be stored,
+    // and for VALUE to be stored correctly it first has to be shifted appropriately.
+    // Note that SHIFT and MASK are always concrete.
+    //
+    // We perform this update in several stages:
+    // 1. First, we simplify MASK &Int #asWord( SLOT ), which results in
+    //    ( VALUE *Int SHIFT ) |Int #asWord ( B1 +Bytes ... +Bytes BN ).
+    // 2. Then, we isolate the +Bytes-junct(s) that will be overwritten.
+    // 3. Then, we write the VALUE, possibly splitting the identified +Bytes-junct(s).
+    //
+    // Note that we require additional simplifications to account for the fact that
+    // VALUE and SLOT can also be concrete. In the former case, we need to extract the
+    // SHIFT appropriate, and in the latter case, the slot will appear on the LHS of the |Int.
+
+    // 1. Slot masking using &Int
+    rule [mask-b-and]:
+      MASK:Int &Int SLOT:Int =>
+        #asWord ( #buf ( 32, SLOT ) [ 32 -Int ( #getMaskShiftBytes(MASK) +Int #getMaskWidthBytes(MASK) ) := #buf ( #getMaskWidthBytes(MASK), 0 ) ] )
+        requires #rangeUInt(256, MASK) andBool #rangeUInt(256, SLOT)
+        andBool #isMask(MASK)
+        [simplification, concrete(MASK), preserves-definedness]
```

Comment:
I guess `#rangeUInt(256, MASK)` is redundant with `#isMask(MASK)`? No problem leaving it in if there is some reason, even readability, but checking my understanding.

Reply:
Yes, it is redundant.

Reply:
Done.